### PR TITLE
Update to 1.0

### PR DESCRIPTION
**Pull request by weymouth**
Upgraded the solver for backend agnostic execution:

- New version of `@loop` macro which integrates KernelAbstractions.jl (KA) to run multi-threaded on CPUs and GPUs. This replaces the `@simd` version of `@loop` as well as previous multi-threading code. Each `@loop <expr> over <I in R>` is expanded into a `@kernel function` and then run on the backend of the first variable in `<expr>`. 
-  BREAKING CHANGE: Many high-level functions don't compile or run correctly or run much slower than expected on GPUs. Things as simple as `sum` or `LinearAlgebra:norm2`. These have been replaced in the code-base with lower-level functions, but unfortunately, users will need to take extra care when defining things like `AutoBody(sdf, map)` functions.
- PERFORMANCE NOTE: KA allocates to the CPU on every loop. Reverting `@loop` to use `@simd` restores a perfectly non-allocating `sim_step!`. We tried other tools like `Polyester.jl` which had better multi-threading performance for small simulations, but large simulations is where we need the speed-up and so we chose KA.
- PERFORMANCE NOTE: `@loop` is not fully optimized. For example, there is an execution overhead for each `@loop` call on GPUs. A few of the loops have been combined to help reduce this overhead, but many more would require major refactoring or modification of the `@loop` macro. Despite this we [benchmarked up to 182x speed-up with GPU execution](https://arxiv.org/abs/2304.08159).
- BREAKING CHANGE: The `Simulation` constructor arguments have changed. `dims` is now the _internal_ field dimension `(L,2L)` not `(L+2,2L+2)`, and `U` must now be an `NTuple`.
- The `Simulation` constructor also take a new `mem=Array` argument which can be set to `CUDA.CuArray` or `AMDGPU.ROCArray` to set-up simulations on GPUs. The `Flow` and `Poisson` structs now use `AbstractArrays` for all fields to accommodate those arrays types.
- DEFAULT CHANGE: `sim_step!(remeasure=true)` is now the default as that is the safer (but slower) option. 
- `Poission` now shares memory for the `L`, `x`, and `z` fields with `Flow` to reduce the memory footprint. The `z` field holds the RHS vector and is mutated by `solve!`.
- The `SOR!` and `GS!` smoothers are not thread-safe, and have been replaced with a Jacobi preconditioned conjugate-gradient smoother held in new routines `Jacobi!` and `pcg!`. 
- PERFORMANCE NOTE: Because of the poor-scaling on small fields, the number of multi-grid levels has been set to a default `maxlevels=4`. The optimal number of levels is likely to be simulation and backend dependent.
- PERFORMANCE NOTE: `pcg!` requires _a lot_ of inner products, which are somewhat slow. Switching to the [data-driven approximate inverse smoother](https://github.com/weymouth/DataDrivenGMG.jl) may be beneficial in the future.
- Because of the poor-scaling on small fields, the multi-grid-style recursive `apply_sdf!` has been replaced with `measure_sdf!` which simply `@loop`s `body.sdf()`.

There have also many changes to the code outside of `src` to support the upgrade:

- The testing cases have been massively expanded. In particular, there are tests for every major function on CPU, CUDA, and AMDGPU backends. 
- The benchmarks have been massively expanded. In particular, benchmarks for each function within `mom_step!` as well as the 3D TGV and donut cases can be compared against previous commits, including pre 1.0 versions.
- The examples have been brought up-to-date, including GPU execution for the 3D examples and a new [jelly fish example](examples/ThreeD_jelly.jl) demonstrating a deforming geometry.

The only (intentional) modelling change was to add  `correct_div!(σ)` to `Body.jl` to enable the deformable jelly fish example. This has nothing to do with the backend upgrade and should have been added to master and then merged in - but it wasn't.

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -93,50 +139,66 @@ end
...
-    @inside p[I] = cos(atan(reverse(loc(0,I) .- 2^6)...))
-    @test sum(abs2,WaterLily.∮nds(p,body)/(π*2^5).-(1,0))<1e-6
+    J = CartesianIndex(2,3,4)
+    for f ∈ arrays
+        u = zeros(3,4,5,3) |> f; apply!((i,x)->x[i]+prod(x),u)
+        p = zeros(3,4,5) |> f
+        @inside p[I] = WaterLily.ke(I,u)
+        @test @allowscalar p[J]==0.5*(26^2+27^2+28^2)
+        @inside p[I] = WaterLily.ke(I,u,(2,3,4))
+        @test @allowscalar p[J]===1.5*24^2
```
Comment (b-fg) [inline test/runtests.jl:172]: `===` should be `==`?

File: benchmark/donut/donut.jl [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+#     # Taylor-Green-Vortex initial velocity field
+#     function uλ(i,vx)
+#         x,y,z = @. (vx-1.5)*π/L                # scaled coordinates
+#         i==1 && return -U*sin(x)*cos(y)*cos(z) # u_x
+#         i==2 && return  U*cos(x)*sin(y)*cos(z) # u_y
+#         return 0.                              # u_z
+#     end
+#     # Initialize simulation
+#     return Simulation((L, L, L), (0, 0, 0), L; U=U, uλ=uλ, ν=ν, T=T, mem=backend)
+# end
```
Comment (b-fg) [inline benchmark/donut/donut.jl:30]: all this get removed

File: test/runtests.jl (hunk 1 above)
Comment (weymouth) [reply test/runtests.jl]: fixed

File: benchmark/donut/donut.jl (hunk 2 above)
Comment (weymouth) [reply benchmark/donut/donut.jl]: deleted

**Pull request merged**